### ISSUE TITLE
CUDA unified memory fail to increase on demand in r1.10 on PPC64le

### ISSUE THREAD

**Issue by alapha23** [tensorflow/tensorflow#21600]
### Description

Based on the commit to [introduce an option to allocate CUDA unified memory](https://github.com/tensorflow/tensorflow/commit/b1139814f91c5216eb5ff229ee7e1982e5f4e888?diff=split), theoretically we could expect a memory swap between CPU and GPU while running datasets with high memory demands. 

However, my tensorflow build still hits OOM. 
I was training [official resnet with cifar10 dataset](https://github.com/tensorflow/models/tree/master/official/resnet), with also a large batch ---- 48*128. 

I built another version without CUDA unified memory from commit [fixed ppc64le compile failure libpng](https://github.com/tensorflow/tensorflow/commit/5aefa441276b5fdf2fec5e7cb282630c104f6f4a), 
Both tensorflow build hits OOM when I try to increase batch size from 47\*128 to 48\*128. 
Based on this comparision, I was convinced that CUDA unified memory is not functioning in my build. 

However, I tried to set `per_process_gpu_memory_fraction` to 200 and pass the config to the `tf.Session`, it worked for much larger batches. 
Then I realized that the UVM cannot grow on demand. Setting `allow_growth` to True won't help. 

In a word, can we make the CUDA unified memory allocation grow on demand?

Could you possibly suggest a reason?
Did I miss out any configurations?

Thank you in advance!
----------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:  no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  Linux Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**:  source
- **TensorFlow version (use command below)**:  r1.10
- **Python version**:  3.5.2
- **Bazel version (if compiling from source)**:  0.15.0
- **GCC/Compiler version (if compiling from source)**:  gcc (Ubuntu/IBM 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
- **CUDA/cuDNN version**:  9.2/7.1
- **GPU model and memory**:   4 x Tesla P100, each has 16G memory; 1T RAM

**Comment by alapha23** [comment]
@smit-hinsu 
Thank you for the commit [ introduce an option to allocate CUDA unified memory](https://github.com/tensorflow/tensorflow/commit/b1139814f91c5216eb5ff229ee7e1982e5f4e888?diff=split)! 

Could you possibly suggest a reason?
Thank you in advance!

**Comment by byronyi** [comment]
You could try to switch to PoolAllocator if you suspect it's the problem of BFCAllocator.

**Comment by byronyi** [comment]
As @zheng-xq have mentioned, TF has systematically switched to BFCAllocator since TF 1.10. You could checked out v1.9 and [here] (https://github.com/tensorflow/tensorflow/blob/r1.9/tensorflow/core/common_runtime/gpu/process_state.cc#L190) to see how to reenable PoolAllocator in TF with UVM.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @smit-hinsu: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @smit-hinsu: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @smit-hinsu: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by smit-hinsu** [comment]
If I understand the experiment correctly, you are seeing that total memory usage does not go above per_process_gpu_memory_fraction with UVM. Is that correct? That seems to be working as intended and consistent with the mode without UVM.

could you describe your use-case requiring total memory to go above per_process_gpu_memory_fraction? wouldn't it be enough to set per_process_gpu_memory_fraction to a very high number and enable allow_growth option?

**Comment by alapha23** [comment]
@smit-hinsu 
Thank you for your reply. 

True. Total memory usage does not go above per_process_gpu_memory_fraction with UVM.

In my experiments, it would be faster when `per_process_gpu_memory_fraction` is smaller. 
Setting `per_process_gpu_memory_fraction` to a very high number would make training much much slower which is not applicable. 

That means I have to find the best `per_process_gpu_memory_fraction` with many times of trying and failing. 

Can you possibly make the total memory allocation(GPU+CPU) grow automatically when we need more memory?

In addition, could you possibly suggest:
- Which memory(CPU or GPU memory) would be first used when we set `per_process_gpu_memory_fraction`  to a high number?
    - As I recognized, gpu memory would not be fully utilized when we set `per_process_gpu_memory_fraction`  to a high number
- What is `allow_growth` doing?

Thank you in advance!

**Comment by smit-hinsu** [comment]
Did you try setting per_process_gpu_memory_fraction to a high number along with enabling allow_growth option?

It should make the total memory allocation(GPU+CPU) to grow automatically when more memory is required as you requested. See allow_growth documentation at https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/config.proto#L36.

Memory is first allocated on CPU and migrated to GPU on demand. It is expected that training will become slower with unified memory as size of the working memory increases. It would generate many faults and stall the training process until memory is moved to GPU.

See cuda unified memory documentation for more details:
https://devblogs.nvidia.com/maximizing-unified-memory-performance-cuda/
http://on-demand.gputechconf.com/gtc/2017/presentation/s7285-nikolay-sakharnykh-unified-memory-on-pascal-and-volta.pdf

Hope this helps!

**Comment by alapha23** [comment]
@smit-hinsu 
Problem solved. Thank you very much!
